rectools/models/sasrec.py

```diff
@@ -932,125 +934,71 @@ def _collate_fn_train(
+        Parameters
+        ----------
+        batch: tp.Dict[str, torch.Tensor]
+            Full batch from train_dataloader. Different aspects (like x or y) can be accessed with corresponding keys.
+        item_embs: torch.Tensor
+            Full catalog item embeddings from torch_model.forward method.
+        session_embs: torch.Tensor
+            Session embeddings from torch_model.forward method.
+        """
+        logits = self._get_logits(batch, item_embs, session_embs)
+        return self._calc_loss(batch, logits)
 
-    def predict_step(self, batch: Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
-        """Prediction step."""
+    def _get_logits(
+        self, batch: tp.Dict[str, torch.Tensor], item_embs: torch.Tensor, session_embs: torch.Tensor
+    ) -> torch.Tensor:
         raise NotImplementedError()
 
+    def _calc_loss(self, batch: tp.Dict[str, torch.Tensor], logits: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError()
 
-class SessionEncoderLightningModule(SessionEncoderLightningModuleBase):
-    """Lightning module to train SASRec model."""
-
-    def on_train_start(self) -> None:
-        """Initialize parameters with values from Xavier normal distribution."""
-        # TODO: init padding embedding with zeros
-        self._xavier_normal_init()
-
-    def training_step(self, batch: Dict[str, torch.Tensor], batch_idx: int) -> torch.Tensor:
-        """TODO"""
-        x, y, w = batch["x"], batch["y"], batch["yw"]
-        if self.loss == "softmax":
-            logits = self._get_full_catalog_logits(x)
-            return self._calc_softmax_loss(logits, y, w)
-        if self.loss == "BCE":
-            negatives = batch["negatives"]
-            logits = self._get_pos_neg_logits(x, y, negatives)
-            return self._calc_bce_loss(logits, y, w)
-        if self.loss == "gBCE":
-            negatives = batch["negatives"]
-            logits = self._get_pos_neg_logits(x, y, negatives)
-            return self._calc_gbce_loss(logits, y, w, negatives)
-
-        raise ValueError(f"loss {self.loss} is not supported")
-
-    def _get_full_catalog_logits(self, x: torch.Tensor) -> torch.Tensor:
-        item_embs, session_embs = self.torch_model(x)
-        logits = session_embs @ item_embs.T
-        return logits
-
-    def _get_pos_neg_logits(self, x: torch.Tensor, y: torch.Tensor, negatives: torch.Tensor) -> torch.Tensor:
-        # [n_items + n_item_extra_tokens, n_factors], [batch_size, session_max_len, n_factors]
-        item_embs, session_embs = self.torch_model(x)
-        pos_neg = torch.cat([y.unsqueeze(-1), negatives], dim=-1)  # [batch_size, session_max_len, n_negatives + 1]
-        pos_neg_embs = item_embs[pos_neg]  # [batch_size, session_max_len, n_negatives + 1, n_factors]
-        # [batch_size, session_max_len, n_negatives + 1]
-        logits = (pos_neg_embs @ session_embs.unsqueeze(-1)).squeeze(-1)
-        return logits
+    @property
```

Comment:
Why is it a property? Does it depend on a specific instance? For me it looks like a class attribute

Reply:
If I'm right, please also use `tp.ClassVar` in the type annotation (just found it recently)